pkg/kubelet/winstats/winstats.go

```diff
@@ -0,0 +1,287 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +build windows
+
+package winstats
+
+import (
+	"context"
+	"encoding/json"
+	dockerstatstypes "github.com/docker/docker/api/types"
+	dockerapi "github.com/docker/engine-api/client"
+	dockertypes "github.com/docker/engine-api/types"
+	cadvisorapi "github.com/google/cadvisor/info/v1"
+	cadvisorapiv2 "github.com/google/cadvisor/info/v2"
+	"k8s.io/kubernetes/pkg/kubelet/network"
+	"os"
+	"runtime"
+	"sync"
+	"time"
+)
+
+type Client struct {
+	dockerClient                *dockerapi.Client
+	cpuUsageCoreNanoSeconds     uint64
+	memoryPrivWorkingSetBytes   uint64
+	memoryCommitedBytes         uint64
+	mu                          sync.Mutex
+	memoryPhysicalCapacityBytes uint64
+}
+
+func NewClient() (*Client, error) {
+	client := new(Client)
+
+	dockerClient, _ := dockerapi.NewEnvClient()
+	client.dockerClient = dockerClient
+
+	// create physical memory
+	memory, err := getPhysicallyInstalledSystemMemoryBytes()
+
+	if err != nil {
+		return nil, err
+	}
+
+	client.memoryPhysicalCapacityBytes = memory
+
+	// start node monitoring (reading perf counters)
+	errChan := make(chan error, 1)
+	go client.startNodeMonitoring(errChan)
+
+	err = <-errChan
+	return client, err
+}
+
+func (c *Client) startNodeMonitoring(errChan chan error) {
+	cpuChan, err := readPerformanceCounter(CPUQuery, 1)
+
+	if err != nil {
+		errChan <- err
+		return
+	}
+
+	memWorkingSetChan, err := readPerformanceCounter(MemoryPrivWorkingSetQuery, 1)
+
+	if err != nil {
+		errChan <- err
+		return
+	}
+
+	memCommittedBytesChan, err := readPerformanceCounter(MemoryCommittedBytesQuery, 1)
+
+	if err != nil {
+		errChan <- err
+		return
+	}
+
+	// no error, send nil over channel
+	errChan <- nil
+
+	for {
+		select {
+		case cpu := <-cpuChan:
+			c.mu.Lock()
+			cpuCores := runtime.NumCPU()
+
+			// This converts perf counter data which is cpu percentage for all cores into nanoseconds.
+			// The formula is (cpuPercentage / 100.0) * #cores * 1e+9 (nano seconds). More info here:
+			// https://github.com/kubernetes/heapster/issues/650
+
+			c.cpuUsageCoreNanoSeconds += uint64((cpu.Value / 100.0) * float64(cpuCores) * 1000000000)
+
+			c.mu.Unlock()
+		case mWorkingSet := <-memWorkingSetChan:
+			c.mu.Lock()
+			c.memoryPrivWorkingSetBytes = uint64(mWorkingSet.Value)
+			c.mu.Unlock()
+		case mCommitedBytes := <-memCommittedBytesChan:
+			c.mu.Lock()
+			c.memoryCommitedBytes = uint64(mCommitedBytes.Value)
+			c.mu.Unlock()
+		}
+	}
+}
+
+func (c *Client) WinContainerInfos() (map[string]cadvisorapiv2.ContainerInfo, error) {
+	infos := make(map[string]cadvisorapiv2.ContainerInfo)
+
+	// root (node) container
+	infos["/"] = *c.createRootContainerInfo()
+
+	containers, err := c.dockerClient.ContainerList(context.Background(), dockertypes.ContainerListOptions{})
+
+	if err != nil {
+		return nil, err
+	}
+
+	for _, container := range containers {
```

Comment:
I would use index ranging here. Using instance ranging will copy the `type.Container` every time which is expensive.  Instead;

```go
for ix := range containers {
   container = &containers[ix]
   ...
}
```

i know it kind of is unfortunate that Go works this way...